buildSrc/src/main/java/org/elasticsearch/gradle/internal/test/rerun/executer/RerunTestResultProcessor.java

```diff
@@ -0,0 +1,121 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.gradle.api.tasks.testing.TestResult.ResultType.SKIPPED;
+
+final class RerunTestResultProcessor implements TestResultProcessor {
+
+    private final TestResultProcessor delegate;
+
+    private final Map<Object, TestDescriptorInternal> activeDescriptorsById = new HashMap<>();
+
+    private TestNames currentRoundFailedTests = new TestNames();
+    private TestNames previousRoundFailedTests = new TestNames();
+
+    private Object rootTestDescriptorId;
+
+    private List<Map<Object,TestDescriptorInternal>> storedActiveDescriptorsWhenFailed = new ArrayList<>();
+
+    RerunTestResultProcessor(TestResultProcessor delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void started(TestDescriptorInternal descriptor, TestStartEvent testStartEvent) {
+        if (rootTestDescriptorId == null) {
+            rootTestDescriptorId = descriptor.getId();
+            activeDescriptorsById.put(descriptor.getId(), descriptor);
+            delegate.started(descriptor, testStartEvent);
+        } else if (descriptor.getId().equals(rootTestDescriptorId) == false) {
+            activeDescriptorsById.put(descriptor.getId(), descriptor);
+            delegate.started(descriptor, testStartEvent);
+        }
+    }
+
+    @Override
+    public void completed(Object testId, TestCompleteEvent testCompleteEvent) {
+        if (testId.equals(rootTestDescriptorId)) {
+            if (lastRun() == false) {
+                return;
+            }
+        } else {
+            TestDescriptorInternal descriptor = activeDescriptorsById.remove(testId);
+            if (descriptor != null && descriptor.getClassName() != null) {
+                String className = descriptor.getClassName();
+                String name = descriptor.getName();
+                boolean failedInPreviousRound = previousRoundFailedTests.remove(className, name);
+                if (failedInPreviousRound && testCompleteEvent.getResultType() == SKIPPED) {
+                    currentRoundFailedTests.add(className, name);
+                }
+            }
+        }
+
+        delegate.completed(testId, testCompleteEvent);
+    }
+
+    @Override
+    public void output(Object testId, TestOutputEvent testOutputEvent) {
+        delegate.output(testId, testOutputEvent);
+    }
+
+    @Override
+    public void failure(Object testId, Throwable throwable) {
+        final TestDescriptorInternal descriptor = activeDescriptorsById.get(testId);
+        if (descriptor != null) {
+            String className = descriptor.getClassName();
+            if (className != null) {
+                currentRoundFailedTests.add(className, descriptor.getName());
+            }
+        }
+
+        delegate.failure(testId, throwable);
+    }
+
+    private boolean lastRun() {
+        return activeDescriptorsById.size() == 1;
+    }
+
+    public RoundResult getResult() {
```

Comment:
This doesn't seem to be used?

Reply:
> Also, what does this look like in build scans if we have partial test execution an then retry? Do test get reported as flaky as they do with the Gradle test retry plugin?

It depends on the result of the tests. 
1. If all tests pass in the first run and in the 2nd run (not taking the system exit causing one into account) then tests are only shown as one time executed in build scan (see example of the build scan I added to the infra test of this PR: https://gradle-enterprise.elastic.co/s/wsczh3rwdcvq4/tests) 
2. If a test failed e.g. in the first execution and then succeed in the 2nd (again not taking the system exit causing one) then we see those tests marked as flaky (see https://gradle-enterprise.elastic.co/s/6vli7zag4ra7u/tests)

Reply:
removed